packages/lesswrong/server/callbacks/compileEbook.ts

```diff
@@ -0,0 +1,308 @@
+//         });
+//       })
+//       .catch(err => {
+//         console.log(err)
+//       })  
+//   }
+// }
+
+const getConfigFromSequence = (sequence: DbSequence, author: DbUser, imagePath: string) => {
+  return {
+      id: slugify(sequence.title),
+      title: sequence.title,
+      cover: imagePath,
+      series: siteNameWithArticleSetting.get(),
+      sequence: 1,
+      author: author.displayName,
+      fileAs: author.displayName,
+      // TODO come back to
+      // genre: "Non-Fiction",
+      // tags: "Rationality",
+      // copyright: `${author.displayName}, `, "Scott Alexander, 2017",
+      publisher: siteNameWithArticleSetting.get(),
+      // published: "2017",
+      language: "en",
+      description: "",
+      contents: "Chapters",
+      source: getSiteUrl(),
+      images: []
+  }
+}
+
+const htmlTemplate = `
+<?xml version="1.0" encoding="utf-8"?>
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
+  <head>
+    <meta http-equiv="Content-Type" content="application/xhtml+xml; charset=utf-8" />
+    <link rel="stylesheet" href="style/base.css" />
+  </head>
+<body></body>
+</html>
+`
+
+export async function ChaptersEditEbookCallback (chapter: DbChapter) {
+  console.log("BEGIN ChaptersEditEbookCallback")
+  const context = await createAdminContext();
+  const posts = await sequenceGetAllPosts(chapter.sequenceId, context)
+  const chapters = await Chapters.find({sequenceId: chapter.sequenceId}).fetch()
+  const sequence = await Sequences.findOne({_id: chapter.sequenceId})
+  const author = await Users.findOne({_id: sequence?.userId})
+
+  if (!sequence) throw Error("No sequence found")
+  if (!author) throw Error("No author found")
+
+  // get cover image
+  const imageUrl = makeCloudinaryImageUrl(sequence.bannerImageId, {f:'jpg'})
+  const file = fs.createWriteStream("tmp/ebookImage0.jpg")
+  
+  https.get(imageUrl, (response) => {
+    response.pipe(file);
+    file.on("finish", async () => {
+      const config = getConfigFromSequence(sequence, author, '/Users/raymondarnold/Documents/LessWrongSuite/Lesswrong2/tmp/ebookImage0.jpg')
+      let epub = makepub.document(config);
+    
+      // epub.addCSS(jetpack.read('style/base.css')); 
+      epub.addSection('Title Page', "<h1>[[TITLE]]</h1><h3>by [[AUTHOR]]</h3>", true, true);
+      
+      function buildEbookFromSequence(sequence, chapters, posts) {
+        // 
+        createDocFromLWContents(sequence.title, sequence)
+        for (let chapter of chapters) {
+          if (chapter.title) {
+            createDocFromLWContents(chapter.title, chapter)
+          }
+          for (let post of posts) {
+            createDocFromLWContents(post.title, post)
+          }
+        }
+      }
+      
+      function createDocFromLWContents(title, document) {
+        let newDoc = cheerio.load(htmlTemplate)
```

Comment:
IMO you probably want to be setting up a React renderer here, similar to how email rendering works, rather than using `cheerio`. I'm not sure whether you can use the same set of components as the main site (emails don't), but even if you can't, you'll be much happier writing this in React than writing it in DOM manipulations.